rfcs/017-fair-share-scheduling.md

```diff
@@ -0,0 +1,216 @@
+
+ Every `UniformPriorityBoostInterval` seconds (called `t2` in the paper):
+   For each `Job`:
+     ```
+     DecayCoeff = get_uniform_decay_coefficient(0.75),
+     Job#job.priority = Job#job.priority * DecayCoeff
+     ```
+
+[note]: If jobs were scheduled to run at an absolute future time (a deadline) this step could be avoided. Then, the effect of all the jobs needing to periodically move to the front of the queue would be accomplished instead by the current time (i.e. `now()`) moving head along the time-line.
+
+3) Adjust running process priority in proportion to the shares used by all the
+jobs in the same db:
+
+ Every `RunningPriorityReduceInterval` seconds (called `t3` in the paper):
+   For each `Job`:
+     ```
+     Db = Job#job.db,
+     SharesSq = shares(Db) * shares(Db),
+     Job#job.priority = Job#job.priority + (usage(Db) * pending(Db)) / SharesSq
+     ```
+
+### How Jobs Start and Stop
+
+During each rescheduling cycle, `max_churn` running jobs from the back of the
+queue are stopped and `max_churn` jobs from the front of the pending queue are
+started. This part is not modified from the existing scheduling algorithm,
+except now, the jobs would be ordered by their `priority` value before being
+ordered by their last start time.
+
+In addition, `one-shot` replication jobs would still be skipped when stopping
+and we'd let them run in order to maintain traditional replication semantics
+just like before.
+
+When picking the jobs to run exclude jobs which have been exponentially backed
+off due to repeated errors. This part is unmodified and from the original
+scheduler.
+
+### Configuration
+
+The decay coefficients and interval times for each of the 3 parts of the algorithm would be configurable in the `[replicator]` config section.
+
+Per-`_replicator` db shares would be configurable in the `[replicator.shares]` section as:
+
+```
+[replicator.shares]
+$prefix/_replicator = $numshares
+```
+
+By default each db is assigned 100 shares. Then higher number of shares should
+then indicated a larger proportion of scheduler resources allocated to that db.
+A lower number would get proportionally less shares.
+
+For example:
+
+```
+[replicator.shares]
+
+; This is the default
+; _replicator = 100
+
+high/_replicator = 200
+low/_replicator = 50
+```
+
+# Advantages and Disadvantages
+
+Advantages:
+
+  * Allow a fair share of resources between multiple `_replicator` db instances
+
+  * Can boost or lower the priority of some replication jobs by adjusting the
+    shares assigned to that database instance.
+
+Disadvantages:
+
+  * Adds more complexity to the scheduler
+
+# Key Changes
+
+ * Modifies replication scheduler
+
+   each `_replicator` db in the system.
+
+ * A delay in `running` state as reflected in monitoring API responses
+
+ * `[replicator] update_docs = false` configuration option becomes hard-coded
+
+## Applications and Modules affected
+
+ * `couch_replicator` application
+
+## HTTP API additions
+
+N/A
```

Comment:
Do you think some API changes would be useful so that people could get info about and monitor replications and compare that with the configuration? Such as: this is the target share distribution; this is how much of that is "utilized".
I'm thinking of basically any ops-aid.

Reply:
Not a bad idea. There is a balance here with adding new HTTP API to 3.x, but I can see it being helpful since the algorithm here is dynamic and it may not be obvious what's happening. Am thinking maybe something like:

```json
   "_scheduler/share_stats" : {
     "_replicator": {
         "shares": 100,
         "utilization": 0.5,
         "pending_jobs": 100,
         "running_jobs": 3
      },
      "high/_replicator": {
         "shares": 200,
         "utilization": 0.8,
         "pending_jobs": 48,
         "running_jobs": 12
    }
```

However a slight problem here that _scheduler/ API works at the cluster level but schedulers work independently on each node. We can aggregate stats which sum easily (for example, if we had `GET _scheduler` return a job total summary by state `{"running": 10, "pending": 90, ...}`)  some of the share stats might not work well aggregated, say if they are normalized values like utilization. We could still sum them I suppose document the fact that they represent a sum. Another option is to return per-node stats here.